DRA: generated resource claim names

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

One concern raised during the initial DRA reviews and again in https://github.com/kubernetes/kubernetes/pull/116254#discussion_r1134010324 was the risk of name conflicts between automatically and manually created ResourceClaims.

This PR addresses that by switching to generated names for automatically created ResourceClaims, with a pod.status extension to record that generated name for code where listing all ResourceClaims is impossible (kubelet) or more complex (node authorizer).

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/113722

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
The names of ResourceClaims generated from ResourceClaimTemplate are now generated. The base name is still `<pod>-<claim name>`, but a random suffix will avoid name collisions.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
dra: generated name for ResourceClaim from template

Generating the name avoids all potential name collisions. It's not clear how
much of a problem that was because users can avoid them and the deterministic
names for generic ephemeral volumes have not led to reports from users. But
using generated names is not too hard either.

What makes it relatively easy is that the new pod.status.resourceClaimStatus
map stores the generated name for kubelet and node authorizer, i.e. the
information in the pod is sufficient to determine the name of the
ResourceClaim.

The resource claim controller becomes a bit more complex and now needs
permission to modify the pod status. The new failure scenario of "ResourceClaim
created, updating pod status fails" is handled with the help of a new special
"resource.kubernetes.io/pod-claim-name" annotation that together with the owner
reference identifies exactly for what a ResourceClaim was generated, so
updating the pod status can be retried for existing ResourceClaims.

The transition from deterministic names is handled with a special case for that
recovery code path: a ResourceClaim with no annotation and a name that follows
the Kubernetes <= 1.27 naming pattern is assumed to be generated for that pod
claim and gets added to the pod status.

There's no immediate need for it, but just in case that it may become relevant,
the name of the generated ResourceClaim may also be left unset to record that
no claim was needed. Components processing such a pod can skip whatever they
normally would do for the claim. To ensure that they do and also cover other
cases properly ("no known field is set", "must check ownership"),
resourceclaim.Name gets extended.

pohly committed:
dra: generated files

fedebongio commented:
/remove-sig api-machinery

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3200,6 +3200,19 @@ type ClaimSource struct {
+// the corresponding ResourceClaim.
+type PodResourceClaimStatus struct {
+	// Name uniquely identifies this resource claim inside the pod.
+	// This must be a DNS_LABEL.
```
And must match a defined resource claim?

pohly replied:
Yes. I wonder whether that should replace "must be a DNS_LABEL" because that follows from "must be the name of an entry in pod.spec.volumeClaims". I'll use both for now.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3200,6 +3200,19 @@ type ClaimSource struct {
+	Name string
+
+	// ResourceClaimName is the name of the ResourceClaim that was
+	// generated for the Pod.
```
...in the same namespace as this pod. 

?

pohly replied:
Right, let's call that out explicitly.

thockin reviewed (commented):
Also need to fix comments in `type ClaimSource`  - "The name of the ResourceClaim will be..."

pohly commented:
This PR is ready for review, the implementation should be complete.

bart0sh commented:
/triage accepted
/priority important-soon

ibihim commented:
From sig-auth perspective, the RBAC change is fine (structurally fine).
Sig-Node needs to review that the RBAC makes sense.

pohly commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3524,22 +3524,31 @@ type ClaimSource struct {
+	// Name uniquely identifies this resource claim inside the pod.
+	// This must match the name of an entry in pod.spec.resourceClaims,
+	// which implies that the string must be a DNS_LABEL.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
```
See https://github.com/kubernetes/kubernetes/pull/117351/commits/5d8727241d9dec2def88648223823b2901a28f4b for a revision of this API which should work for ResourceClaims that were generated for a network.

pohly replied:
I wasn't happy with https://github.com/kubernetes/kubernetes/commit/5d8727241d9dec2def88648223823b2901a28f4b. If we ever do something for networks, let's add a `pod.status.networkStatuses` that code can check to find a generated ResourceClaim.

The part about better error handling in `resourceclaim.Name` callers and allowing ResourceClaim creation to be skipped still made sense to me, so I kept that in the latest revision of this PR, after squashing and rebasing (because of a conflict).

pohly commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4174,6 +4194,12 @@ type PodStatus struct {
+	// +listType=atomic
+	// +featureGate=DynamicResourceAllocation
+	// +optional
+	ResourceClaimStatuses []PodResourceClaimStatus `json:"resourceClaimStatuses,omitempty" protobuf:"bytes,15,rep,name=resourceClaimStatuses"`
```
This cannot be `+listType:map` because the field that could serve as key, the `sourceRef`, is a struct and those cannot be keys.

It also cannot be a `listType=set` because the strategic merging cannot be used to patch just this one field. The apiserver then preserves old and new entries (I tried...).

This is okay. There's a single writer (the controller) and it can ensure that the field gets updated atomically.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/resourceclaim.go:
```diff
@@ -24,32 +24,66 @@ limitations under the License.
-func Name(pod *v1.Pod, podClaim *v1.PodResourceClaim) string {
-	if podClaim.Source.ResourceClaimName != nil {
-		return *podClaim.Source.ResourceClaimName
+func Name(pod *v1.Pod, podClaim *v1.PodResourceClaim) (name *string, mustCheckOwner bool, err error) {
```
I used the opportunity to tighten validation in clients (controller manager, kubelet) which need to look up claims. This revised implementation of `Name` implements the "Consumers of this type must treat an empty object as if it has an unknown value." part of the API description.

This was partly left to the caller before and not always done consistently.

klueska replied:
How much overhead is it to just always check for the owner? It just feels awkward to have this extra boolean returned from here. Or better yet, why can't we just check for the owner in here and return an error if the pod passed in is not the owner whenever it "must" be checked.

pohly replied:
Not all claims will have the pod as owner. Only those generated for the pod are required to have that.

The code returning the boolean doesn't have access to the claim object and therefore cannot check it. The usual pattern is "determine claim name, get claim, check that it is usable" - the latter then includes checking the ownership.

klueska replied:
OK, that makes sense with not having enough information to do the check from within the call, but `mustCheckOwner ` still feels out of place as a return value to a method called `Name()`.  Can we make it its own API call somehow?

klueska replied:
OK. I don't feel too strongly, so we can leave it as is.

pohly commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -314,82 +334,204 @@ func (ec *Controller) syncPod(ctx context.Context, namespace, name string) error
+	}
+
 	templateName := podClaim.Source.ResourceClaimTemplateName
 	if templateName == nil {
```
This is where previously, the controller would have returned without an error when `podClaim.Source` was empty (= future unknown field set).

Treating this as an error (new code) is consistent with the API description and while it may be too strict, it's probably safer because if a new field gets added that has to be handled by the kube-controller-manager, then running with an older release will fail, as it should in this case.

pohly commented:
The more I think about it, the less happy I am with https://github.com/kubernetes/kubernetes/pull/117351/commits/5d8727241d9dec2def88648223823b2901a28f4b. The goal was to make `pod.status.resourceClaimStatuses` future-proof and to support storing the name of a ResourceClaim that was generated for a network.

But that goal can also be achieved by adding a new `pod.status.networkStatuses` field if and when needed. The code would be simpler.

What makes sense IMHO is to support "ResourceClaim not generated". The API remains sane (`ResourceClaimName *string` instead of `ResoureClaimName string`) and perhaps we'll need it at some point.

pohly committed:
dra e2e: check that not generating a ResourceClaim works

This is not something that normally happens, but the API supports it because it
might be needed at some point, so we have to test it.

pohly commented:
/hold cancel

liggitt commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -314,82 +336,179 @@ func (ec *Controller) syncPod(ctx context.Context, namespace, name string) error
 		}
 	}
 
+	if newPodClaims != nil {
```
this means we're now doing two writes in a single run of the control loop (create of the resourceclaim and update of the status)... doesn't this introduce new error cases, like:
* resourceclaim creation succeeds, pod status update fails, resourceclaim gets orphaned
* controller resourceclaim cache doesn't yet contain a previously created resourceclaim object for the pod, so we double create with a new generated name

pohly replied:
I've added one commit which adds a `MutationCache`. Wasn't too complex and thus probably is worth doing.

liggitt commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -314,82 +336,179 @@ func (ec *Controller) syncPod(ctx context.Context, namespace, name string) error
+		claimName := *claimName
+		// The ResourceClaim should exist because it is recorded in the pod.status.resourceClaimStatuses,
+		// but perhaps it was deleted accidentally. In that case we re-create it.
+		claim, err := ec.claimLister.ResourceClaims(pod.Namespace).Get(claimName)
```
this lister could be stale, which would make us think a resourceclaim didn't exist when it really did, and double-create, right?

pohly replied:
Yes - see https://github.com/kubernetes/kubernetes/pull/117351#discussion_r1253197611

pohly committed:
dra: store generated ResourceClaims in cache

This addresses the following bad sequence of events:
- controller creates ResourceClaim
- updating pod status fails
- pod gets retried before the informer receives
  the created ResourceClaim
- another ResourceClaim gets created

Storing the generated ResourceClaim in a MutationCache ensures that the
controller knows about it during the retry.

A positive side effect is that ResourceClaims now get index by pod owner and
thus iterating over existing ones becomes a bit more efficient.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3184,22 +3184,33 @@ type ClaimSource struct {
-	// resource by mistake. Scheduling and pod startup are then blocked
-	// until the unrelated ResourceClaim is removed.
+	// will also be deleted. The ResourceClaim will have a name that is
+	// generated with <pod name>-<resource name> as base and the Pod as
```
You probably want to spec here what happens if the `<pod name>-<resource name>` template is too long to fit in a claim name.

OK to do in a followup

pohly replied:
The concatenated string is stored in `claim.GeneratedName` and then the normal truncation happens when generating the actual name. I started to get worried that validation of `claim.GeneratedName` might reject a too long string, but that doesn't seem to be the case. I added an E2E test with `pod name = strings.Repeat("p", 63)` and `resource name =  strings.Repeat("c", 63)` and that worked:
```
    generateName: ppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppp-ccc
cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
    name: pppppppppppppppppppppppppppppppppppppppppppppppppppppppppptcgdb
```

I'll add a comment about that.

pohly replied:
```
	// The template will be used to create a new ResourceClaim, which will
	// be bound to this pod. When this pod is deleted, the ResourceClaim
	// will also be deleted. The ResourceClaim will have <pod
	// name>-<resource name> as `generateName` base and the Pod as
	// owner. The final name then gets generated by the apiserver (i.e.
	// unique random suffix at the end, if necessary after truncating the
	// base). It gets recorded in pod.status.resourceClaimStatuses.
```

thockin replied:
ok.  The only quesion I have is: GenerateName has very high probability of succeeding at making something random, but in the case of long input, it just truncates.  That's USUALLY acceptable, but the alternative would be for you to do your own truncation, like:

```
if len(podName) + len(resName) > 57 { // 57 + "-<5random>"
    // truncate podName and resName in some more helpful way
}
```

Probably not worth it now, but may consider later.

pohly replied:
If we put "will have <pod name>-<resource name> as `generateName` base" into the API description, can we still change it later without an API break? Clients might have started to rely on exactly that aspect (for whatever reason). Should the exact way of truncating perhaps be left out of the API description after all or be marked as "may change"?

I like doing something better, for example truncating both strings separately. Whatever it is, clients should not rely on the exact process.

thockin replied:
Maybe just spec it as "The pod name and resource name, along with a generated component, will be used to form a unique name for the ResourceClaim, which will be recorded in..."

Also, from a safety point of view, consider what happens if the ResourceClaim is created, and then the controller crashes before recording the resulting name.  Does the ResourceClaim have a label sufficient to re-assemble the state (e.g. a k8s-owned key with the pod's UID as a value)?

pohly replied:
I like that wording better than my current one - changed.

Regarding safety: there is an annotation that allows the controller to continue. See [prior discussion with Jordan](https://github.com/kubernetes/kubernetes/pull/117351#discussion_r1253101683).

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4815,6 +4816,36 @@ func validatePodConditions(conditions []core.PodCondition, fldPath *field.Path)
+		// There's no need to check the content of the name. If it matches an entry,
+		// then it is valid, otherwise we reject it here.
+		if !havePodClaim(podClaims, status.Name) {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), status.Name, "must match the name of an entry in pod.spec.resourceClaims"))
```
we usually denote field names in backticks, and don't name the "this"

```
"must match the name of an entry in `spec.resourceClaims`"
```

pohly replied:
Changed.

thockin reviewed (commented):
I didn't review the controller parts very well - where is it actually composing the string to be used as the Claim name?  I just want to make sure it handles overflow reasonably, so we can spec it.

API LGTM, just nits on comments

thockin commented:
/approve

API approved - I did not review controller - who is reviewing that?

pohly commented:
> I did not review controller - who is reviewing that?

I asked @soltysh today and he agreed to look at it, probably on Monday.

soltysh reviewed (approved):
With comments from Jordan address through use of MutationCache, the controller changes lgtm

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -68,22 +68,35 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string) (
+		klog.V(3).InfoS("Processing resource", "podClaim", podClaim.Name, "pod", pod.Name)
+		claimName, mustCheckOwner, err := resourceclaim.Name(pod, podClaim)
+		if err != nil {
+			return err
```
We should probably wrap this error to be consistent.

pohly replied:
Okay.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -68,22 +68,35 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string) (
+			return err
+		}
 
+		if claimName == nil {
+			// Nothing to do.
+			continue
+		}
```
When would there ever be no error, but `claimName == nil`? I would think that that would be an error condition from within `resourceclaim.Name()`.

klueska replied:
If it's not an error to have a `nil` claim name, then it shouldn't be hidden behind an error. Maybe just adding a comment above the check as to why it might ever be nil is sufficient (at which point we can remove the comment of "Nothing to do" as that should be obvious from the new comment).

pohly replied:
Comment changed.

klueska commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -268,24 +268,24 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		}
+		claim, err := pl.claimLister.ResourceClaims(pod.Namespace).Get(*claimName)
 		if err != nil {
-			// The error usually has already enough context ("resourcevolumeclaim "myclaim" not found"),
-			// but we can do better for generic ephemeral inline volumes where that situation
-			// is normal directly after creating a pod.
-			if isEphemeral && apierrors.IsNotFound(err) {
-				err = fmt.Errorf("waiting for dynamic resource controller to create the resourceclaim %q", claimName)
-			}
```
I don't have a whole lot of context on why this was needed to begin with, but why is this not needed anymore? I guess it's all captured within the `resourceclaim.Name()` call now?

pohly replied:
Correct. It'll check for this based on the pod status field.

klueska replied:
Ack

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -68,22 +68,35 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string) (
+		if err != nil {
+			return fmt.Errorf("prepare resource claim: %v", err)
+		}
 
+		if claimName == nil {
+			// Nothing to do.
+			continue
+		}
```
Can we add a comment such as the following here? Looks like it will need to be propagated to 3 places in total.
```
// The claim name might be nil if no underlying resource claim
// was generated for the referenced claim. There are valid use
// cases when this might happen, so we simply skip it.
```

pohly replied:
[Changed](https://github.com/kubernetes/kubernetes/compare/aff83426e547e83261425b68f50f27356df7a019..fec25785ee6cef3998d2eeb28df068370a9938f2) - please take another look.

klueska commented:
/approve for kubelet changes

klueska commented:
/lgtm
